### PR TITLE
Allow linking against vcomp on Windows

### LINKED ISSUE #54054: Instructions of building PyTorch with Intel OpenMP on Windows
After #52964, I failed to build PyTorch on Windows due to the lacking of intel-openmp. I did my investigation and found the following things:

* `conda install intel-openmp`: installs the `libiomp5md.dll` and related, but NOT headers.
* `conda install mkl mkl-include`: installs the MKL libraries & headers, but NOT anything OpenMP.
* [This link](https://s3.amazonaws.com/ossci-windows/mkl_2020.2.254.7z) used in #52964: include both libraries & headers for both MKL & OpenMP.

So how do we expect users to build the code right now, using intel-openmp? The current instruction no longer works. 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @malfet @seemethere @walterddr @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @mszhanyi @skyline75489

### PR DESCRIPTION
**Pull request by skyline75489**
Fixes #54054

### CODE REVIEW

**Comment by skyline75489** [comment]
Hi @imaginary-person. After some digging this is the solution I think is good enough at the moment.

I found that even oneDNN itself has support for vcomp (See [here](https://github.com/oneapi-src/oneDNN#installation)) . It feels a bit aggressive for us to just drop support for vcomp right now. 

So I relaxed the check in `FindOpenMP.cmake` and added some notes on OpenMP variants. Now users can choose whatever OpenMP runtime they'd like to use, if they had like AMD processors or some other concerns. This does not affect CI build and everything, since those are driven by scripts.

Sadly PRs like #53737 will still break the build. I honestly wish the OpenMP on Windows can keep up the steps with other platforms. But for this moment I think we should *not* merge #53737 and alike, just to offer users more choices.

And FYI Peter has stepped down from the role of maintaining Windows support of PyTorch. I think this PR is also what he wants after carefully reading the related threads.

File: cmake/Modules/FindOpenMP.cmake [hunk 1]
```diff
@@ -101,7 +101,7 @@ function(_OPENMP_FLAG_CANDIDATES LANG)
       set(OMP_FLAG_Intel "-qopenmp")
     endif()
     set(OMP_FLAG_MIPSpro "-mp")
-    set(OMP_FLAG_MSVC "-openmp:experimental" "-openmp:experimental -I${__header_dir}" "-openmp" "-openmp -I${__header_dir}")
+    set(OMP_FLAG_MSVC "-openmp:experimental" "-openmp:experimental -I\"${__header_dir}\"" "-openmp" "-openmp -I\"${__header_dir}\"")
```
Comment (skyline75489) [inline cmake/Modules/FindOpenMP.cmake:104]: This is for paths that has spaces in it, for example `C:\Program Files (x86)\Microsoft Visual Studio\2019\...`

**Review by peterjc123 — CHANGES_REQUESTED** [review]
Apparently, after removing `omp_get_level` in the test sources, the `omp.h` from `VCOMP` is preferred over that from `IntelOMP`, which is unexpected.

**Comment by skyline75489** [comment]
@peterjc123 I think the preferred OpenMP implementation depends on the search paths, right? In the instructions we use ` set "CMAKE_INCLUDE_PATH=%cd%\mkl\include"` to force search `omp.h` provided by Intel first. Why is relaxing the restriction here matters? I don't quite get it.

**Comment by skyline75489** [comment]
I see it in the log. This is using iomp:

```text
-- Found OpenMP_CXX: -openmp:experimental -IC:/Users/circleci/project/build/win_tmp/mkl/include  
```

This is after this PR:

```text
-- Found OpenMP_CXX: -openmp:experimental  
```

So we have no choice but to let vcomp fail, in order to prefer iomp? 😢

**Comment by peterjc123** [comment]
> @peterjc123 I think the preferred OpenMP implementation depends on the search paths, right? In the instructions we use ` set "CMAKE_INCLUDE_PATH=%cd%\mkl\include"` to force search `omp.h` provided by Intel first. Why is relaxing the restriction here matters? I don't quite get it.

`CMAKE_INCLUDE_PATH` is used for searching for the omp header within CMake scripts. You'll still need to modify the search paths by adding the `-I` switch.
Ref: https://cmake.org/cmake/help/latest/variable/CMAKE_INCLUDE_PATH.html

**Comment by imaginary-person** [comment]
BTW, I think `vcomp` was never being linked, but only its `omp.h` was being used by Visual Studio.

**Comment by imaginary-person** [comment]
I see now that the title of the PR is indeed correct. After this PR would land, if a user building from source on Windows wouldn't install Intel OpenMP, [as otherwise specified in the build instructions](https://github.com/pytorch/pytorch/blob/master/docs/source/notes/windows.rst#building-from-source), then `vcomp` would be linked, and `omp.h` provided by `vcomp` would be used.

The following behavior would remain unaffected while building from source on Windows, as desired:
1. If a user would install Intel OpenMP but not MKL, then the `omp.h` header provided by Microsoft would be used, but Intel OpenMP would be linked.
2. If a user would install both Intel OpenMP & MKL via Conda, then the `omp.h` header provided by Intel OpenMP would be used, and Intel OpenMP would be linked.

**Comment by skyline75489** [comment]
@imaginary-person yeah what you said is basically right. Only that you need more than things from Conda. There's no iomp headers in Conda, remember? Even with `conda install intel-openmp`, you'll still need to follow the [instructions](https://github.com/pytorch/pytorch/blob/master/docs/source/notes/windows.rst#building-from-source) written by Peter. The important step is:

```bat
set "CMAKE_INCLUDE_PATH=%cd%\mkl\include"
set "LIB=%cd%\mkl\lib;%LIB%"
```

Only after setting this, CMake will use iomp over vcomp.

You can see this isn't really a smooth experience for users on Windows. This is also the main reason I'd like us to continue support vcomp.

**Comment by imaginary-person** [comment]
@skyline75489, thanks for the clarification! I now understand that before #52964, `FindOpenMP.cmake` was unable to find Intel OpenMP on Windows (non-CI builds) if users didn't specify the OpenMP path manually, so not only was `vcomp`'s `omp.h` being used, but `vcomp` was being linked! I didn't know about this detail earlier.

I agree that this PR would make building from source on Windows more convenient by linking with `vcomp`.

However, talking about the OpenMP library & not the `omp.h` header file, the following command would work for linking Intel OpenMP _only if_ MKL would be downloaded:

> set "LIB=%cd%\mkl\lib;%LIB%"

If only Intel OpenMP is downloaded via Conda, it is installed at a separate location than the one in the above commands (which is for MKL) . Would `FindOpenMP.cmake` be able to find the location of the Intel OpenMP library in this case? If not, the build instructions should be updated so that users who wouldn't install MKL but would want to link the Intel OpenMP library may specify its correct path.

**Comment by skyline75489** [comment]
>  Would FindOpenMP.cmake be able to find the location of the Intel OpenMP library in this case

No, at least based on my experience. 

I added the note about OpenMP is this PR to address this. Do you think it's good enough? Perhaps not 😄

**Comment by imaginary-person** [comment]
> I added the note about OpenMP is this PR to address this. Do you think it's good enough? Perhaps not 😄

Your note is informative, but those [build instructions](https://github.com/pytorch/pytorch/blob/master/docs/source/notes/windows.rst#building-from-source) mention setting the path for MKL lib directory (MKL also installs OpenMP). 

But MKL is an optional component, so if a user would install Intel OpenMP without installing MKL (via another command mentioned in the build instructions), then they should also manually specify its path, but the instructions don't mention doing so. Hence, a user might falsely assume that they've linked Intel OpenMP while their local build would've linked `vcomp`.

**Comment by skyline75489** [comment]
Oh yes. Thanks for mentioning this. I forgot this. I'll try to update the note later.

**Comment by skyline75489** [comment]
OK now it's working properly.

CI builds:

```text
-- Found OpenMP_CXX: -openmp:experimental -IC:/w/b/windows/mkl/include  
```

Local build:

```text
Found OpenMP_CXX: -openmp:experimental
```

**Review by peterjc123 — APPROVED** [review]

**Comment by peterjc123** [comment]
Well, currently, it won't find the `omp.h` header if user installed MKL by themselves, since the include path won't be added as a hint to `find_path`. This needs an additional fix.

**Comment by skyline75489** [comment]
@peterjc123 Do you mean installing MKL using Conda? No OMP header is provided by `mkl-include`, though.

**Comment by peterjc123** [comment]
> @peterjc123 Do you mean installing MKL using Conda? No OMP header is provided by `mkl-include`, though.

I mean installing from the MKL installer. As for `mkl-include`, let's just don't consider that as an option, because you may not construct a MKL development environment that can be constructed by using conda only (omp import library is missing `libiomp5md.lib`)

**Comment by skyline75489** [comment]
OK I understand. I didn't know there's "MKL Installer". Will give it a try later.

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**